File: clients/ios/BuildIt/Features/Social/SocialViewModel.swift
// SocialViewModel.swift
// BuildIt - Decentralized Mesh Communication
//
// ViewModel for social features (feed, posts, reactions).

import Foundation
import Combine
import os.log

/// Post model representing a kind 1 Nostr event
struct SocialPost: Identifiable, Equatable {
    let id: String
    let content: String
    let authorPubkey: String
    var authorName: String?
    var authorAvatar: String?
    let createdAt: Date
    let replyToId: String?
    var reactionCount: Int
    var replyCount: Int
    var userReacted: Bool
    var replies: [SocialPost]

    static func == (lhs: SocialPost, rhs: SocialPost) -> Bool {
        lhs.id == rhs.id
    }
}

/// ViewModel for social features
@MainActor
class SocialViewModel: ObservableObject {
    // MARK: - Published Properties

    @Published var posts: [SocialPost] = []
    @Published var isLoading = false
    @Published var isPosting = false
    @Published var error: String?

    // MARK: - Private Properties

    private let logger = Logger(subsystem: "com.buildit", category: "SocialViewModel")
    private var cancellables = Set<AnyCancellable>()

    // MARK: - Initialization

    init() {
        loadFeed()
    }

    // MARK: - Public Methods

    /// Loads the activity feed from followed contacts
    func loadFeed() {
        Task {
            isLoading = true
            defer { isLoading = false }

            do {
                // Get followed contacts
                let contacts = Database.shared.getAllContacts()
                let pubkeys = contacts.map { $0.publicKey }

                // Subscribe to kind 1 events from contacts
                let events = try await NostrClient.shared.subscribeToKind1(authors: pubkeys)

                // Convert events to posts
                var newPosts: [SocialPost] = []
                for event in events {
                    if let post = await eventToPost(event) {
                        newPosts.append(post)
                    }
                }

                // Sort by date descending
                newPosts.sort { $0.createdAt > $1.createdAt }

                // Keep last 100 posts
                posts = Array(newPosts.prefix(100))

                logger.info("Loaded \(self.posts.count) posts")
            } catch {
                self.error = error.localizedDescription
                logger.error("Failed to load feed: \(error.localizedDescription)")
            }
        }
    }

    /// Creates a new post (NIP-01 kind 1 event)
    func createPost(content: String) {
        Task {
            isPosting = true
            defer { isPosting = false }

            do {
                let event = try await NostrClient.shared.createKind1Event(content: content)
                try await NostrClient.shared.publishEvent(event)

                // Add to local feed
                if let post = await eventToPost(event) {
                    posts.insert(post, at: 0)
                }

                logger.info("Created new post")
            } catch {
                self.error = error.localizedDescription
                logger.error("Failed to create post: \(error.localizedDescription)")
            }
        }
    }

    /// Creates a reply to a post
    func replyToPost(parentId: String, content: String) {
        Task {
            isPosting = true
            defer { isPosting = false }

            do {
                let event = try await NostrClient.shared.createReplyEvent(
                    content: content,
                    replyTo: parentId
                )
                try await NostrClient.shared.publishEvent(event)

                // Reload replies
                await loadReplies(postId: parentId)

                logger.info("Created reply to post \(parentId)")
            } catch {
                self.error = error.localizedDescription
                logger.error("Failed to create reply: \(error.localizedDescription)")
            }
        }
    }

    /// Reacts to a post (NIP-25 kind 7 event)
    func reactToPost(postId: String, emoji: String = "+") {
        Task {
            do {
                let event = try await NostrClient.shared.createReactionEvent(
                    postId: postId,
                    emoji: emoji
                )
                try await NostrClient.shared.publishEvent(event)

                // Update local reaction count
                if let index = posts.firstIndex(where: { $0.id == postId }) {
                    posts[index].reactionCount += 1
                    posts[index].userReacted = true
                }

                logger.info("Reacted to post \(postId)")
            } catch {
                self.error = error.localizedDescription
                logger.error("Failed to react: \(error.localizedDescription)")
            }
        }
    }

    /// Loads replies to a specific post
    func loadReplies(postId: String) async {
        do {
            let replyEvents = try await NostrClient.shared.subscribeToReplies(eventId: postId)

            var replies: [SocialPost] = []
            for event in replyEvents {
                if let reply = await eventToPost(event) {
                    replies.append(reply)
                }
            }

            // Sort by date ascending (oldest first)
            replies.sort { $0.createdAt < $1.createdAt }

            // Update the post's replies
            if let index = posts.firstIndex(where: { $0.id == postId }) {
                posts[index].replies = replies
                posts[index].replyCount = replies.count
            }

            logger.info("Loaded \(replies.count) replies for post \(postId)")
        } catch {
            self.error = error.localizedDescription
            logger.error("Failed to load replies: \(error.localizedDescription)")
        }
    }

    /// Refreshes the feed
    func refresh() {
        posts.removeAll()
        loadFeed()
    }

    /// Clears error state
    func clearError() {
        error = nil
    }

    // MARK: - Private Methods

    /// Converts a Nostr event to a SocialPost
    private func eventToPost(_ event: NostrEvent) async -> SocialPost? {
        guard event.kind == 1 else { return nil }

        let contact = Database.shared.getContact(publicKey: event.pubkey)

        return SocialPost(
            id: event.id,
            content: event.content,
            authorPubkey: event.pubkey,
            authorName: contact?.name,
            authorAvatar: contact?.avatarURL,
            createdAt: Date(timeIntervalSince1970: Double(event.createdAt)),
            replyToId: event.tags.first { $0.first == "e" }?[safe: 1],
            reactionCount: 0,
            replyCount: 0,
            userReacted: false,
            replies: []
        )
    }
}

// MARK: - Array Extension

private extension Array {
    subscript(safe index: Int) -> Element? {
        indices.contains(index) ? self[index] : nil
    }
}

// MARK: - Nostr Event Type (placeholder - should use actual implementation)

struct NostrEvent: Codable {
    let id: String
    let pubkey: String
    let createdAt: Int64
    let kind: Int
    let content: String
    let tags: [[String]]
    let sig: String
}

File: clients/android/app/src/main/java/network/buildit/features/social/SocialViewModel.kt
package network.buildit.features.social

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import network.buildit.core.crypto.CryptoManager
import network.buildit.core.crypto.UnsignedNostrEvent
import network.buildit.core.nostr.NostrClient
import network.buildit.core.nostr.NostrEvent
import network.buildit.core.nostr.NostrFilter
import network.buildit.core.storage.ContactDao
import java.util.UUID
import javax.inject.Inject

/**
 * ViewModel for Social features (activity feed, posts, reactions).
 *
 * Implements NIP-01 kind 1 (text notes) for microblogging.
 */
@HiltViewModel
class SocialViewModel @Inject constructor(
    private val nostrClient: NostrClient,
    private val cryptoManager: CryptoManager,
    private val contactDao: ContactDao
) : ViewModel() {

    private val _uiState = MutableStateFlow(SocialUiState())
    val uiState: StateFlow<SocialUiState> = _uiState.asStateFlow()

    private val _posts = MutableStateFlow<List<Post>>(emptyList())
    val posts: StateFlow<List<Post>> = _posts.asStateFlow()

    init {
        loadFeed()
        collectEvents()
    }

    /**
     * Collects incoming Nostr events and maps kind 1 to posts.
     */
    private fun collectEvents() {
        viewModelScope.launch {
            nostrClient.events.collect { event ->
                if (event.kind == 1) {
                    val post = eventToPost(event)
                    if (post != null) {
                        val currentPosts = _posts.value.toMutableList()
                        if (currentPosts.none { it.id == post.id }) {
                            currentPosts.add(0, post)
                            currentPosts.sortByDescending { it.createdAt }
                            _posts.value = currentPosts.take(100)
                        }
                    }
                }
            }
        }
    }

    /**
     * Loads the activity feed from followed contacts.
     */
    fun loadFeed() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true)

            try {
                // Get followed contacts
                val contacts = contactDao.getAllContacts().first()
                val pubkeys = contacts.map { it.pubkey }

                // Subscribe to kind 1 events from contacts
                nostrClient.subscribe(
                    NostrFilter(
                        authors = pubkeys.ifEmpty { null },
                        kinds = listOf(1),
                        limit = 100
                    )
                )

                _uiState.value = _uiState.value.copy(isLoading = false)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    error = e.message
                )
            }
        }
    }

    /**
     * Creates a new post (NIP-01 kind 1 event).
     */
    fun createPost(content: String) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isPosting = true)

            try {
                val pubkey = cryptoManager.getPublicKeyHex() ?: throw IllegalStateException("No public key")
                val unsigned = UnsignedNostrEvent(
                    pubkey = pubkey,
                    createdAt = System.currentTimeMillis() / 1000,
                    kind = 1,
                    tags = emptyList(),
                    content = content
                )
                val signed = cryptoManager.signEvent(unsigned)
                    ?: throw IllegalStateException("Failed to sign event")

                val event = NostrEvent(
                    id = signed.id,
                    pubkey = signed.pubkey,
                    createdAt = signed.createdAt,
                    kind = signed.kind,
                    tags = signed.tags,
                    content = signed.content,
                    sig = signed.sig
                )
                nostrClient.publishEvent(event)

                // Add to local feed
                val post = eventToPost(event)
                if (post != null) {
                    val currentPosts = _posts.value.toMutableList()
                    currentPosts.add(0, post)
                    _posts.value = currentPosts
                }

                _uiState.value = _uiState.value.copy(isPosting = false)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    isPosting = false,
                    error = e.message
                )
            }
        }
    }

    /**
     * Creates a reply to a post.
     */
    fun replyToPost(parentId: String, content: String) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isPosting = true)

            try {
                val pubkey = cryptoManager.getPublicKeyHex() ?: throw IllegalStateException("No public key")
                val unsigned = UnsignedNostrEvent(
                    pubkey = pubkey,
                    createdAt = System.currentTimeMillis() / 1000,
                    kind = 1,
                    tags = listOf(listOf("e", parentId)),
                    content = content
                )
                val signed = cryptoManager.signEvent(unsigned)
                    ?: throw IllegalStateException("Failed to sign event")

                val event = NostrEvent(
                    id = signed.id,
                    pubkey = signed.pubkey,
                    createdAt = signed.createdAt,
                    kind = signed.kind,
                    tags = signed.tags,
                    content = signed.content,
                    sig = signed.sig
                )
                nostrClient.publishEvent(event)

                // Reload the thread
                loadReplies(parentId)

                _uiState.value = _uiState.value.copy(isPosting = false)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    isPosting = false,
                    error = e.message
                )
            }
        }
    }

    /**
     * Reacts to a post (NIP-25 kind 7 event).
     */
    fun reactToPost(postId: String, emoji: String = "+") {
        viewModelScope.launch {
            try {
                val pubkey = cryptoManager.getPublicKeyHex() ?: throw IllegalStateException("No public key")
                val unsigned = UnsignedNostrEvent(
                    pubkey = pubkey,
                    createdAt = System.currentTimeMillis() / 1000,
                    kind = 7,
                    tags = listOf(listOf("e", postId)),
                    content = emoji
                )
                val signed = cryptoManager.signEvent(unsigned)
                    ?: throw IllegalStateException("Failed to sign event")

                val event = NostrEvent(
                    id = signed.id,
                    pubkey = signed.pubkey,
                    createdAt = signed.createdAt,
                    kind = signed.kind,
                    tags = signed.tags,
                    content = signed.content,
                    sig = signed.sig
                )
                nostrClient.publishEvent(event)

                // Update local reaction count
                val currentPosts = _posts.value.toMutableList()
                val index = currentPosts.indexOfFirst { it.id == postId }
                if (index >= 0) {
                    val post = currentPosts[index]
                    currentPosts[index] = post.copy(
                        reactionCount = post.reactionCount + 1,
                        userReacted = true
                    )
                    _posts.value = currentPosts
                }
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(error = e.message)
            }
        }
    }

    /**
     * Reposts a post (NIP-18 kind 6 repost event).
     */
    fun repostPost(postId: String, authorPubkey: String) {
        viewModelScope.launch {
            try {
                val pubkey = cryptoManager.getPublicKeyHex() ?: throw IllegalStateException("No public key")

                // Find the original post to include its content in the repost
                val originalPost = _posts.value.find { it.id == postId }

                val unsigned = UnsignedNostrEvent(
                    pubkey = pubkey,
                    createdAt = System.currentTimeMillis() / 1000,
                    kind = 6, // NIP-18 repost
                    tags = listOf(
                        listOf("e", postId, ""), // reference to original event
                        listOf("p", authorPubkey) // reference to original author
                    ),
                    content = "" // Content is empty for reposts per NIP-18
                )
                val signed = cryptoManager.signEvent(unsigned)
                    ?: throw IllegalStateException("Failed to sign event")

                val event = NostrEvent(
                    id = signed.id,
                    pubkey = signed.pubkey,
                    createdAt = signed.createdAt,
                    kind = signed.kind,
                    tags = signed.tags,
                    content = signed.content,
                    sig = signed.sig
                )
                nostrClient.publishEvent(event)

                // Update local repost count
                val currentPosts = _posts.value.toMutableList()
                val index = currentPosts.indexOfFirst { it.id == postId }
                if (index >= 0) {
                    val post = currentPosts[index]
                    currentPosts[index] = post.copy(
                        repostCount = post.repostCount + 1,
                        userReposted = true
                    )
                    _posts.value = currentPosts
                }
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(error = e.message)
            }
        }
    }

    /**
     * Loads replies to a specific post.
     */
    fun loadReplies(postId: String) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(loadingReplies = true)

            try {
                // Subscribe to replies (kind 1 with e-tag referencing postId)
                nostrClient.subscribe(
                    NostrFilter(
                        kinds = listOf(1),
                        tags = mapOf("e" to listOf(postId)),
                        limit = 50
                    )
                )

                _uiState.value = _uiState.value.copy(loadingReplies = false)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    loadingReplies = false,
                    error = e.message
                )
            }
        }
    }

    /**
     * Converts a Nostr event to a Post model.
     */
    private suspend fun eventToPost(event: NostrEvent): Post? {
        if (event.kind != 1) return null

        val contact = contactDao.getByPubkey(event.pubkey)

        return Post(
            id = event.id,
            content = event.content,
            authorPubkey = event.pubkey,
            authorName = contact?.displayName,
            authorAvatar = contact?.avatarUrl,
            createdAt = event.createdAt,
            replyToId = event.tags.find { it.firstOrNull() == "e" }?.getOrNull(1),
            reactionCount = 0,
            replyCount = 0,
            userReacted = false,
            replies = emptyList()
        )
    }

    /**
     * Clears any error state.
     */
    fun clearError() {
        _uiState.value = _uiState.value.copy(error = null)
    }

    /**
     * Refreshes the feed.
     */
    fun refresh() {
        _posts.value = emptyList()
        loadFeed()
    }
}

/**
 * UI state for Social features.
 */
data class SocialUiState(
    val isLoading: Boolean = false,
    val isPosting: Boolean = false,
    val loadingReplies: Boolean = false,
    val error: String? = null
)

/**
 * Represents a social post (kind 1 event).
 */
data class Post(
    val id: String,
    val content: String,
    val authorPubkey: String,
    val authorName: String?,
    val authorAvatar: String?,
    val createdAt: Long,
    val replyToId: String?,
    val reactionCount: Int,
    val replyCount: Int,
    val repostCount: Int = 0,
    val userReacted: Boolean,
    val userReposted: Boolean = false,
    val replies: List<Post>
)
